src/allmydata/util/cputhreadpool.py

```diff
@@ -0,0 +1,47 @@
+"""
+A global thread pool for CPU-intensive tasks.
+
+Motivation:
+
+* Certain tasks are blocking on CPU, and so should be run in a thread.
+* The Twisted thread pool is used for operations that don't necessarily block
+  on CPU, like DNS lookups.  CPU processing should not block DNS lookups!
+* The number of threads should be fixed, and tied to the number of available
+  CPUs.
+
+As a first pass, this uses ``os.cpu_count()`` to determine the max number of
+threads.  This may create too many threads, as it doesn't cover things like
+scheduler affinity or cgroups, but that's not the end of the world.
+"""
+
+import os
+from typing import TypeVar, Callable, cast
+from functools import partial
+
+from twisted.python.threadpool import ThreadPool
+from twisted.internet.defer import Deferred
+from twisted.internet.threads import deferToThreadPool
+from twisted.internet.interfaces import IReactorFromThreads
+
+
+_CPU_THREAD_POOL = ThreadPool(minthreads=0, maxthreads=os.cpu_count(), name="TahoeCPU")
+# Daemon threads allow shutdown to happen:
+_CPU_THREAD_POOL.threadFactory = partial(_CPU_THREAD_POOL.threadFactory, daemon=True)
```

Comment:
Did the segfault-on-shutdown bugs in daemon threads get fixed?  Or do we just not care

Reply:
A bunch of them got fixed in 3.3. And there were some regressions(?) in 3.7-3.8-3.9 that got fixed in point releases. (One of) Python 3.8's thread pool implementations used daemon threads, and then they switched to `_atexit_register` mechanism (see below).

Alternatives:

* Register shutdown with a global reactor automatically. Kinda messes with tests...
* Can't use `atexit.register` cause thread pools prevent exit.
* There's `threading._register_atexit`... which is private, but available on CPython/PyPy 3.9+, and specifically designed for thread pools.
* Make it a Service.

So maybe... use `_register_atexit` if possible, otherwise daemon?